File: sections/61_icpc_contest_readiness/problems/c_first_round_assignment/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    teammates, cards, edges = data[:3]
    graph = [[] for _ in range(teammates + 1)]
    index = 3
    for _ in range(edges):
        teammate, card = data[index], data[index + 1]
        index += 2
        graph[teammate].append(card)

    match_card = [0] * (cards + 1)

    def augment(teammate: int) -> bool:
        for card in graph[teammate]:
            if seen[card]:
                continue
            seen[card] = True
            if match_card[card] == 0 or augment(match_card[card]):
                match_card[card] = teammate
                return True
        return False

    answer = 0
    for teammate in range(1, teammates + 1):
        seen = [False] * (cards + 1)
        answer += augment(teammate)
    print(answer)


if __name__ == "__main__":
    main()

File: sections/61_icpc_contest_readiness/problems/c_first_round_assignment/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int teammates, cards, edges;
    if (!(cin >> teammates >> cards >> edges)) return 0;
    vector<vector<int>> graph(teammates + 1);
    while (edges--) {
        int teammate, card;
        cin >> teammate >> card;
        graph[teammate].push_back(card);
    }

    vector<int> match_card(cards + 1);
    vector<char> seen(cards + 1);
    function<bool(int)> augment = [&](int teammate) {
        for (int card : graph[teammate]) {
            if (seen[card]) continue;
            seen[card] = true;
            if (match_card[card] == 0 || augment(match_card[card])) {
                match_card[card] = teammate;
                return true;
            }
        }
        return false;
    };

    int answer = 0;
    for (int teammate = 1; teammate <= teammates; ++teammate) {
        fill(seen.begin(), seen.end(), false);
        answer += augment(teammate);
    }
    cout << answer << '\n';
}
